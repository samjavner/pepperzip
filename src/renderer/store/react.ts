import * as React from "react";
import { Commands, Dispatch, Store, Update, UseStoreOptions } from "./";

export interface Msg<M extends object, U extends Update<M>> {
    type: keyof U;
    payload?: any;
}

export function useLocalStore(log: (msg: any) => void = () => ({})) {
    return function actual<
        M extends object,
        U extends Update<M>,
        C extends Commands = {}
    >({
        init,
        update,
        createCommands,
        memo = [],
    }: UseStoreOptions<M, U, C>): Store<M, U, C> {
        const [model, reducerDispatch] = React.useReducer(
            (state: M, action: Msg<M, U>) =>
                update[action.type](state, action.payload),
            init
        );

        const vm = React.useMemo(
            () => {
                const dispatch: Dispatch<U> = {} as any;
                Object.keys(update).forEach(type => {
                    (dispatch as any)[type] = (payload: any) => {
                        const msg =
                            update[type].length === 1
                                ? { type }
                                : { type, payload };
                        log(msg);
                        reducerDispatch(msg);
                    };
                });
                const commands: C = createCommands
                    ? createCommands(model, dispatch)
                    : (({} as any) as C);
                return {
                    ...model,
                    ...dispatch,
                    ...commands,
                };
            },
            [model, update, ...memo]
        );

        return vm;
    };
}
