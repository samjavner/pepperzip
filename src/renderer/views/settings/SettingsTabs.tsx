import { Button, ButtonGroup } from "@blueprintjs/core";
import classNames from "classnames";
import * as React from "react";
import { connect } from "react-redux";
import { Dispatch } from "redux";
import * as actions from "../../app/actions";
import { State } from "../../app/state";

function mapStateToProps(state: State) {
    return {
        activePage: state.activeSettingsPage,
    };
}

function mapDispatchToProps(dispatch: Dispatch<actions.Action>) {
    return {
        onSettingsClick: () =>
            dispatch(actions.setActiveSettingsPage("settings")),
        onAboutClick: () => dispatch(actions.setActiveSettingsPage("about")),
    };
}

type SettingsTabsProps = ReturnType<typeof mapStateToProps> &
    ReturnType<typeof mapDispatchToProps>;

export const SettingsTabs: React.SFC<SettingsTabsProps> = props => (
    <>
        <div className="pz-navtabs">
            <ButtonGroup minimal={true}>
                <Button
                    className={classNames({
                        "pz-navtabs-button": true,
                        "pz-navtabs-button--active":
                            props.activePage === "settings",
                    })}
                    icon="cog"
                    onClick={props.onSettingsClick}
                >
                    Settings
                </Button>
                <Button
                    className={classNames({
                        "pz-navtabs-button": true,
                        "pz-navtabs-button--active":
                            props.activePage === "about",
                    })}
                    icon="info-sign"
                    onClick={props.onAboutClick}
                >
                    About
                </Button>
            </ButtonGroup>
        </div>
        <div style={{ borderTop: "1px solid lightgray" }} />
    </>
);

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(SettingsTabs);
