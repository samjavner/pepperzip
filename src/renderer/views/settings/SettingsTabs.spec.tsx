import * as React from "react";
import * as renderer from "react-test-renderer";
import { SettingsTabs } from "./SettingsTabs";

describe("SettingsTabs", () => {
    const onSettingsClick = () => ({} as any);
    const onAboutClick = () => ({} as any);

    it("should render with the 'settings' tab active when the 'settings' page is active", () => {
        const component = renderer.create(
            <SettingsTabs
                activePage="settings"
                onSettingsClick={onSettingsClick}
                onAboutClick={onAboutClick}
            />
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render with the 'about' tab active when the 'about' page is active", () => {
        const component = renderer.create(
            <SettingsTabs
                activePage="about"
                onSettingsClick={onSettingsClick}
                onAboutClick={onAboutClick}
            />
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should handle click on 'settings' tab", () => {
        const component = renderer.create(
            <SettingsTabs
                activePage="settings"
                onSettingsClick={onSettingsClick}
                onAboutClick={onAboutClick}
            />
        );
        const button = component.root.find(
            node =>
                node.props.className &&
                node.props.className.indexOf("pz-navtabs-button--active") !== -1
        );
        expect(button.props.onClick).toBe(onSettingsClick);
    });

    it("should handle click on 'about' tab", () => {
        const component = renderer.create(
            <SettingsTabs
                activePage="about"
                onSettingsClick={onSettingsClick}
                onAboutClick={onAboutClick}
            />
        );
        const button = component.root.find(
            node =>
                node.props.className &&
                node.props.className.indexOf("pz-navtabs-button--active") !== -1
        );
        expect(button.props.onClick).toBe(onAboutClick);
    });
});
