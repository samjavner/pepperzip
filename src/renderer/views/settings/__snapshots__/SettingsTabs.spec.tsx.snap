// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsTabs should render with the 'about' tab active when the 'about' page is active 1`] = `
Array [
  <div
    className="pz-navtabs"
  >
    <div
      className="bp3-button-group bp3-minimal"
    >
      <button
        className="bp3-button pz-navtabs-button"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        type="button"
      >
        <span
          className="bp3-icon bp3-icon-cog"
          icon="cog"
        >
          <svg
            data-icon="cog"
            height={16}
            viewBox="0 0 16 16"
            width={16}
          >
            <desc>
              cog
            </desc>
            <path
              d="M15.19 6.39h-1.85c-.11-.37-.27-.71-.45-1.04l1.36-1.36c.31-.31.31-.82 0-1.13l-1.13-1.13a.803.803 0 0 0-1.13 0l-1.36 1.36c-.33-.17-.67-.33-1.04-.44V.79c0-.44-.36-.8-.8-.8h-1.6c-.44 0-.8.36-.8.8v1.86c-.39.12-.75.28-1.1.47l-1.3-1.3c-.3-.3-.79-.3-1.09 0L1.82 2.91c-.3.3-.3.79 0 1.09l1.3 1.3c-.2.34-.36.7-.48 1.09H.79c-.44 0-.8.36-.8.8v1.6c0 .44.36.8.8.8h1.85c.11.37.27.71.45 1.04l-1.36 1.36c-.31.31-.31.82 0 1.13l1.13 1.13c.31.31.82.31 1.13 0l1.36-1.36c.33.18.67.33 1.04.44v1.86c0 .44.36.8.8.8h1.6c.44 0 .8-.36.8-.8v-1.86c.39-.12.75-.28 1.1-.47l1.3 1.3c.3.3.79.3 1.09 0l1.09-1.09c.3-.3.3-.79 0-1.09l-1.3-1.3c.19-.35.36-.71.48-1.1h1.85c.44 0 .8-.36.8-.8v-1.6a.816.816 0 0 0-.81-.79zm-7.2 4.6c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3z"
              fillRule="evenodd"
            />
          </svg>
        </span>
        <span
          className="bp3-button-text"
        >
          Settings
        </span>
      </button>
      <button
        className="bp3-button pz-navtabs-button pz-navtabs-button--active"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        type="button"
      >
        <span
          className="bp3-icon bp3-icon-info-sign"
          icon="info-sign"
        >
          <svg
            data-icon="info-sign"
            height={16}
            viewBox="0 0 16 16"
            width={16}
          >
            <desc>
              info-sign
            </desc>
            <path
              d="M8 0C3.58 0 0 3.58 0 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zM7 3h2v2H7V3zm3 10H6v-1h1V7H6V6h3v6h1v1z"
              fillRule="evenodd"
            />
          </svg>
        </span>
        <span
          className="bp3-button-text"
        >
          About
        </span>
      </button>
    </div>
  </div>,
  <div
    style={
      Object {
        "borderTop": "1px solid lightgray",
      }
    }
  />,
]
`;

exports[`SettingsTabs should render with the 'settings' tab active when the 'settings' page is active 1`] = `
Array [
  <div
    className="pz-navtabs"
  >
    <div
      className="bp3-button-group bp3-minimal"
    >
      <button
        className="bp3-button pz-navtabs-button pz-navtabs-button--active"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        type="button"
      >
        <span
          className="bp3-icon bp3-icon-cog"
          icon="cog"
        >
          <svg
            data-icon="cog"
            height={16}
            viewBox="0 0 16 16"
            width={16}
          >
            <desc>
              cog
            </desc>
            <path
              d="M15.19 6.39h-1.85c-.11-.37-.27-.71-.45-1.04l1.36-1.36c.31-.31.31-.82 0-1.13l-1.13-1.13a.803.803 0 0 0-1.13 0l-1.36 1.36c-.33-.17-.67-.33-1.04-.44V.79c0-.44-.36-.8-.8-.8h-1.6c-.44 0-.8.36-.8.8v1.86c-.39.12-.75.28-1.1.47l-1.3-1.3c-.3-.3-.79-.3-1.09 0L1.82 2.91c-.3.3-.3.79 0 1.09l1.3 1.3c-.2.34-.36.7-.48 1.09H.79c-.44 0-.8.36-.8.8v1.6c0 .44.36.8.8.8h1.85c.11.37.27.71.45 1.04l-1.36 1.36c-.31.31-.31.82 0 1.13l1.13 1.13c.31.31.82.31 1.13 0l1.36-1.36c.33.18.67.33 1.04.44v1.86c0 .44.36.8.8.8h1.6c.44 0 .8-.36.8-.8v-1.86c.39-.12.75-.28 1.1-.47l1.3 1.3c.3.3.79.3 1.09 0l1.09-1.09c.3-.3.3-.79 0-1.09l-1.3-1.3c.19-.35.36-.71.48-1.1h1.85c.44 0 .8-.36.8-.8v-1.6a.816.816 0 0 0-.81-.79zm-7.2 4.6c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3z"
              fillRule="evenodd"
            />
          </svg>
        </span>
        <span
          className="bp3-button-text"
        >
          Settings
        </span>
      </button>
      <button
        className="bp3-button pz-navtabs-button"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        type="button"
      >
        <span
          className="bp3-icon bp3-icon-info-sign"
          icon="info-sign"
        >
          <svg
            data-icon="info-sign"
            height={16}
            viewBox="0 0 16 16"
            width={16}
          >
            <desc>
              info-sign
            </desc>
            <path
              d="M8 0C3.58 0 0 3.58 0 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zM7 3h2v2H7V3zm3 10H6v-1h1V7H6V6h3v6h1v1z"
              fillRule="evenodd"
            />
          </svg>
        </span>
        <span
          className="bp3-button-text"
        >
          About
        </span>
      </button>
    </div>
  </div>,
  <div
    style={
      Object {
        "borderTop": "1px solid lightgray",
      }
    }
  />,
]
`;
